examples/cpp/delivery_mechanisms/PubSubApp.cpp

```diff
@@ -0,0 +1,340 @@
+// Copyright 2024 Proyectos y Sistemas de Mantenimiento SL (eProsima).
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @file PubSubApp.cpp
+ *
+ */
+
+#include "PubSubApp.hpp"
+
+#include <condition_variable>
+#include <stdexcept>
+
+#include <fastdds/dds/core/status/SubscriptionMatchedStatus.hpp>
+#include <fastdds/dds/domain/DomainParticipantFactory.hpp>
+#include <fastdds/dds/publisher/DataWriter.hpp>
+#include <fastdds/dds/publisher/Publisher.hpp>
+#include <fastdds/dds/publisher/qos/DataWriterQos.hpp>
+#include <fastdds/dds/publisher/qos/PublisherQos.hpp>
+#include <fastdds/dds/subscriber/DataReader.hpp>
+#include <fastdds/dds/subscriber/qos/DataReaderQos.hpp>
+#include <fastdds/dds/subscriber/qos/SubscriberQos.hpp>
+#include <fastdds/dds/subscriber/SampleInfo.hpp>
+#include <fastdds/dds/subscriber/Subscriber.hpp>
+#include <fastdds/rtps/transport/shared_mem/SharedMemTransportDescriptor.h>
+#include <fastdds/rtps/transport/TCPv4TransportDescriptor.h>
+#include <fastdds/rtps/transport/UDPv4TransportDescriptor.h>
+
+#include "Application.hpp"
+#include "CLIParser.hpp"
+#include "DeliveryMechanismsPubSubTypes.h"
+
+using namespace eprosima::fastdds::dds;
+using namespace eprosima::fastdds::rtps;
+
+namespace eprosima {
+namespace fastdds {
+namespace examples {
+namespace delivery_mechanisms {
+
+PubSubApp::PubSubApp(
+        const CLIParser::entity_config& config,
+        const std::string& topic_name)
+    : participant_(nullptr)
+    , publisher_(nullptr)
+    , subscriber_(nullptr)
+    , topic_(nullptr)
+    , reader_(nullptr)
+    , writer_(nullptr)
+    , type_(new DeliveryMechanismsPubSubType())
+    , matched_(0)
+    , index_of_last_sample_sent_(0)
+    , received_samples_(0)
+    , samples_(config.samples)
+    , stop_(false)
+{
+    // Create the participant
+    DomainParticipantQos pqos = PARTICIPANT_QOS_DEFAULT;
+    pqos.name("DeliveryMechanisms_pubsub_participant");
+    pqos.transport().use_builtin_transports = false;
+
+    uint32_t max_samples = samples_;
+    if (max_samples == 0)
+    {
+        max_samples = DATAWRITER_QOS_DEFAULT.resource_limits().max_samples_per_instance;
+    }
+    // Special definitions for certain delivery mechanisms
+    std::shared_ptr<SharedMemTransportDescriptor> shm_transport_ = std::make_shared<SharedMemTransportDescriptor>();
```

Comment:
I would rearrange this part as follow:

* In my opinion the default behaviour in case of `config.delivery_mechanism = DeliveryMechanismKind::DEFAULT` should be the builtin_transport so I would set it to true (or not setting anything since it's the default value).
* Declare all the transport shared pointer inside the switch if they are actually used. 
* Unify this switch with the one used to define the data_sharing. Create a data_sharing_ variable set to true or false inside the switch.

Here how I would rearrange it:
```    LibrarySettings library_settings;
    library_settings.intraprocess_delivery = IntraprocessDeliveryType::INTRAPROCESS_OFF;

    if (config.ignore_local_endpoints)
    {
        pqos.properties().properties().emplace_back(
            "fastdds.ignore_local_endpoints",
            "true");
    }
    bool data_sharing_ = false;
    switch (config.delivery_mechanism)
    {
        case CLIParser::DeliveryMechanismKind::INTRA_PROCESS:   // (It should never reach this section)
            {
            // No transport needed, but at least a transport needs to be declared to avoid participant creation failure
            pqos.transport().use_builtin_transports = true;
            library_settings.intraprocess_delivery = IntraprocessDeliveryType::INTRAPROCESS_FULL;
            break;
            }
        case CLIParser::DeliveryMechanismKind::SHM:
            {
            pqos.transport().use_builtin_transports = false;
            std::shared_ptr<SharedMemTransportDescriptor> shm_transport_ = std::make_shared<SharedMemTransportDescriptor>();
            shm_transport_->segment_size(shm_transport_->max_message_size() * max_samples);
            pqos.transport().user_transports.push_back(shm_transport_);
            break;
            }
        case CLIParser::DeliveryMechanismKind::DATA_SHARING:
            {
            std::shared_ptr<SharedMemTransportDescriptor> shm_transport_ = std::make_shared<SharedMemTransportDescriptor>();
            shm_transport_->segment_size(shm_transport_->max_message_size() * max_samples);
            pqos.transport().use_builtin_transports = false;
            pqos.transport().user_transports.push_back(shm_transport_);
            data_sharing_ = true;
            break;
            }
        case CLIParser::DeliveryMechanismKind::TCP:
            {
            pqos.transport().use_builtin_transports = false;
            pqos.wire_protocol().builtin.discovery_config.leaseDuration = eprosima::fastrtps::c_TimeInfinite;
            pqos.wire_protocol().builtin.discovery_config.leaseDuration_announcementperiod = Duration_t(5, 0);
            Locator tcp_initial_peers_locator_;
            tcp_initial_peers_locator_.kind = LOCATOR_KIND_TCPv4;
            tcp_initial_peers_locator_.port = 0;
            eprosima::fastrtps::rtps::IPLocator::setIPv4(tcp_initial_peers_locator_, "127.0.0.1");
            std::shared_ptr<TCPv4TransportDescriptor> tcp_transport_ = std::make_shared<TCPv4TransportDescriptor>();
            tcp_transport_->sendBufferSize = 0;
            tcp_transport_->receiveBufferSize = 0;
            tcp_transport_->set_WAN_address("127.0.0.1");
            tcp_transport_->add_listener_port(5100);
            pqos.transport().user_transports.push_back(tcp_transport_);
            break;
            }
        case CLIParser::DeliveryMechanismKind::UDP:
            {
            pqos.transport().use_builtin_transports = false;
            pqos.transport().user_transports.push_back(std::make_shared<UDPv4TransportDescriptor>());
            break;
            }
        default:
            pqos.transport().use_builtin_transports = true;
            break;
    }
   
```

Reply:
I've applied all these suggestions but with a different approach:
- I am not setting the ignore local endpoints in the publisher and the subscriber because it does not make sense
- Instead of creating  a new data-sharing variable and adding an if-else sentence below, I compare directly in the if-else the delivery mechanism variable